Add MethodBase.IsConstructedGenericMethod property

DawidTabak opened the pull request:
Added IsConstructedGenericMethod property to MethodBase, for completion and expresiveness as described in #17051 

Closes #17051

danmosemsft commented:
@AtsushiKan are the tests sufficient?

@Petermarcu this would expose an API earlier implemented in CLR. It's a convenience API. Would you like to hold on merging this til after 5/10?

DawidTabak commented:
I see my build fails when using -platform=netfx but succeeds with -platform=netcoreapp
The IsConstructedGenericMethod property I've added to MethodBase can't be found when running with -platform=netfx.

Could someone point me in the right direction with this? 
Am I missing an reference somewhere?

karelz commented on src/System.Reflection.Emit/ref/System.Reflection.Emit.cs:
```diff
@@ -231,6 +231,7 @@ public sealed partial class MethodBuilder : System.Reflection.MethodInfo
         public bool InitLocals { get { throw null; } set { } }
         public override bool IsGenericMethod { get { throw null; } }
         public override bool IsGenericMethodDefinition { get { throw null; } }
+        public override bool IsConstructedGenericMethod { get { throw null; } }
```
You either need to implement this one in the src project (in CoreCLR I assume), or remove the ref. @AtsushiKan @DnlHarvey will be able to provide guidance ...

AtsushiKan replied:
It should be implemented in the source (CoreCLR).

This method has a default implementation which should suffice so you might have to do anything here, unless one of our api compat checking tools complain about an override method missing in the implementation.

karelz commented:
Also regarding your additional questions in https://github.com/dotnet/corefx/issues/17051#issuecomment-298170136

> I think I'm supposed to replace the MethodBase.IsConstructedGenericMethod implementation in ref project with an throw null, if I understand the ref project correctly.

Correct

> I believe the same throw null is needed in MethodBuilder in System.Reflection.Emit
Could someone please confirm this? 

I don't think so (as stated above), but @AtsushiKan @DnlHarvey will be able to say for sure.

> I also see that the IsConstructedGenericMethod of TypeBuilder is not tested in System.Reflection.Emit tests. 
Should the IsConstructedGenericMethod of MethodBuilder be tested or should I omit the tests as it is omitted in TypeBuilder tests?

Please add tests. Pre-existing test gaps in an area should never be an excuse for not adding tests for new APIs :). If you can actually add tests also for the pre-existing test-gap, that would be highly appreciated. Thank you!

Also note that we are finishing .NET Core 2.0 -- we will delay any new APIs and larger changes which are not critical for 2.0 until after May 10 (1.5 weeks). Just setting the expectation ... see https://github.com/dotnet/corefx/issues/17619#issuecomment-296872132 for details.
I will mark this issue as such, unless @AtsushiKan thinks this is something we should push into 2.0 (which is IMO unlikely).

Thanks for your contribution!

AtsushiKan commented:
>I see my build fails when using -platform=netfx 

This api does not exist on netfx.

I would add a IsConstructedGenericMethodSupported property to PlatformDetection.cs (you can use <a href="https://github.com/dotnet/corefx/blob/master/src/Common/tests/System/PlatformDetection.cs#L360">IsNonZeroLowerBoundArraySupported</a> as a template) and guard the tests that way.

AtsushiKan commented on src/System.Runtime/tests/System/Reflection/MethodBaseTests.cs:
```diff
@@ -23,19 +23,34 @@ public static void Test_GetCurrentMethod()
         }
 
         [Fact]
+        public static void Test_GetCurrentMethod_GenericConstructedMethod()
```
I would name this test "Test_GetCurrentMethod_ConstructedGenericMethod" so the names line up with the api naming.

AtsushiKan commented on src/System.Runtime/tests/System/Reflection/MethodBaseTests.cs:
```diff
@@ -23,19 +23,34 @@ public static void Test_GetCurrentMethod()
+            Assert.False(m.IsGenericMethodDefinition);
+            Assert.True(m.IsConstructedGenericMethod);
+            Assert.Equal(1, m.GetGenericArguments().Length);
+            Assert.Equal("Byte", m.GetGenericArguments()[0].Name);
```
I would compare to "typeof(byte)" rather than comparing the Name.

AtsushiKan commented on src/System.Runtime/tests/System/Reflection/MethodBaseTests.cs:
```diff
@@ -23,19 +23,34 @@ public static void Test_GetCurrentMethod()
+            MethodBase m = MyFakeGenericConstructedMethod<Byte>();
+
+            Assert.Equal("MyFakeGenericConstructedMethod", m.Name);
+            Assert.Equal("MethodBaseTests", m.ReflectedType.Name);
```
You can compare to typeof(MethodBaseTests) rather than comparing type names.

AtsushiKan commented on src/System.Runtime/tests/System/Reflection/MethodBaseTests.cs:
```diff
@@ -98,6 +113,13 @@ public static MethodBase MyFakeGenericMethod<T>()
 
+        public static MethodBase MyFakeGenericConstructedMethod<T>()
+        {
+            MethodBase m = MethodBase.GetCurrentMethod();
```
I would avoid the use of GetCurrentMethod(). You can retrieve generic methods on types using any of the usual Type.GetMethod() overloads.

DawidTabak commented:
>> I see my build fails when using -platform=netfx

>This api does not exist on netfx.
>I would add a IsConstructedGenericMethodSupported property to PlatformDetection.cs (you can use >IsNonZeroLowerBoundArraySupported as a template) and guard the tests that way.

@AtsushiKan 
I've added the guard in PlatformDetection.cs where I check MethodInfo if it contains the IsConstructedGenericMethod property. I initially tried to simply call the IsConstructedGenericMethod getter, however that failed because netfx MethodBase does not contain the 'IsConstructedGenericMethod' definition. 

I can add the PlatformDetection.IsConstructedGenericMethodSupported guards in test code, however the compile time issue of missing IsConstructedGenericMethod definition on netfx remains.

Which, I think, means one of the following is needed:
- IsConstructedGenericMethod implementation in netfx which throws the PlatformNotSupportedException.
- Use of preprocessor directives to skip the IsConstructedGenericMethod code when on netfx.
- Use of reflection to retrieve and call the 'IsConstructedGenericMethod' property on MethodBase instance, which avoids the compile time issue.

Currently I've implemented the third option. Is this correct, or should I take other approach. (I'm quite new to netfx)

AtsushiKan commented:
>I see my build fails when using -platform=netfx

Ah, I did not realize this would fail at build time on netfx. I though we would have been building against the same reference assemblies and failing with a MissingMethodException at runtime. Given that that's not true, PlatformDetection is not a useful way to solve this. Sorry for sending you down that path.

The best way to handle this then is to put the new tests in a .cs file that's conditionally included in the build for netcore configurations. There's already such a file: <a href="https://github.com/dotnet/corefx/blob/master/src/System.Runtime/tests/System.Runtime.Tests.csproj#L124">TypeTests.netcoreapp.cs</a>. You can create a new one ("MethodBaseTests.netcoreapp.cs") and include it within the same ItemGroup as TypeTests.netcoreapp.cs (the ItemGroup that conditions on TargetGroup being netcoreapp or uapaot) Then your new test will be excluded from the build on netfx builds and you can write the test the straightforward way (no PlatformDetection checks, no Reflection.)

karelz commented:
@DawidTabak thanks a lot for your contribution!
I have also updated our docs-under-construction with hint about the .netcoreapp.cs test files - see [wiki docs](https://github.com/dotnet/corefx/wiki/Contributions#expose-api-across-coreclrcorefx)
